tools/sof-fw-boot-test.py

```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import subprocess
+import sys
+import time
+from pathlib import Path
+
+# define command line arguments
+def parse_cmdline():
+	parser = argparse.ArgumentParser(add_help=True, formatter_class=argparse.RawTextHelpFormatter,
+	    description='A script for stress testing firmware boot')
+
+	parser.add_argument('-i', '--iter', type=int, default=100, help="number of firmware boot iterations")
+	parser.add_argument('-f', '--firmware', type=str, required=True, help='firmware filename')
+	parser.add_argument('-p', '--fw_path', type=str, required=True, help='path to the firmware file relative to /lib/firmware')
+
+	return vars(parser.parse_args())
+
+def main():
+	cmd_args = parse_cmdline()
+
+	# Get firmware file path
+	fw_path = cmd_args["fw_path"]
+	full_fw_path = '/lib/firmware/' + fw_path
+	if not os.path.exists(full_fw_path):
+		print(f"Firmware path {fw_path} does not exist")
+		sys.exit(1)
+
+	# Get firmware file name
+	fw_filename = cmd_args["firmware"]
+
+	# check if the firmware binary exists
+	fw_fullname = '/lib/firmware/' + fw_path + '/' + fw_filename
+	if not os.path.isfile(fw_fullname):
+		print(f"Firmware file {fw_filename} does not exist")
+		sys.exit(1)
+
+	num_iter = cmd_args["iter"]
+	output = f"""==============================================================================
+	Starting boot stress test with:
+	Firmware filename: {fw_filename}
+	Path to firmware file: {fw_path}
+	Number of Iterations: {num_iter}
+=============================================================================="""
+	print(output)
+
+	# set path to the depbugfs entries
+	debugfs_path = '/sys/kernel/debug/sof/dsp_ops'
+	if not os.path.exists(debugfs_path):
+		print(f"Debugfs path {debugfs_path} does not exist!")
+		sys.exit(1)
+
+	# set the firmware filename & path
+	os.system('echo %s > %s/fw_filename' %(fw_filename, debugfs_path))
+	os.system('echo %s > %s/fw_path' %(fw_path, debugfs_path))
+	total_boot_time_ms = 0
+	min_boot_time_ms = sys.maxsize
+	max_boot_time_ms = 0
+
+	for i in range(num_iter):
+		# put the DSP in D3 with a timeout of 1 second
+		debugfs_entry = debugfs_path + '/dsp_power_state'
+		cmd = 'echo D3 > ' + debugfs_entry
+		subprocess.run(cmd, shell=True, check=True, timeout=1)
```

Comment:
Can this really time out? If not, see above.

If it can, you can still to Python and save some duplication like this (TESTED)

Again: more useful stack trace and simpler debugging when it times out (or fails in any way).

```python

import multiprocessing

def _write_func(fdest, string):
    with open(fdest, 'w') as f:
        f.write(string)

def write_timeout(fdest, string, timeout):
    p = multiprocessing.Process(target=_write_func, args=[fdest, string])
    p.start()
    p.join(timeout)
    if p.exitcode is None:
        raise multiprocessing.TimeoutError

write_timeout(debugfs_entry, 'D3', timeout=2)
```

Reply:
i've dropped the timeout checks. Realistically, I dont think this will timeout. If fw boot times out in the kernel, it will return an error anyway. So that should be enough